File: .NET/Microsoft.Recognizers.Text.NumberWithUnit/Utilities/DictionaryUtils.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;

namespace Microsoft.Recognizers.Text.NumberWithUnit.Utilities
{
    public static class DictionaryUtils
    {
        // Safely bind dictionary which contains several key-value pairs to the destination dictionary.
        // This function is used to bind all the prefix and suffix for units.
        public static void BindDictionary(IDictionary<string, string> dictionary, IDictionary<string, string> sourceDictionary)
        {
            if (dictionary == null)
            {
                return;
            }

            foreach (var pair in dictionary)
            {
                if (string.IsNullOrEmpty(pair.Key))
                {
                    continue;
                }

                BindUnitsString(sourceDictionary, pair.Key, pair.Value);
            }
        }

        // Bind keys in a string which contains words separated by '|'.
        public static void BindUnitsString(IDictionary<string, string> sourceDictionary, string key, string source)
        {
            var values = source.Trim().Split('|');

            foreach (var token in values)
            {
                if (string.IsNullOrWhiteSpace(token) || (sourceDictionary.ContainsKey(token) && sourceDictionary[token].Equals(key, StringComparison.Ordinal)))
                {
                    continue;
                }

                // This segment of code is going to break if there're duplicated key-values in the resource files.
                // Those duplicates should be fixed before committing.
                try
                {
                    sourceDictionary.Add(token, key);
                }
                catch (ArgumentException ae)
                {
                    throw new ArgumentException(ae.Message + ": " + token);
                }
            }
        }
    }
}

File: Java/libraries/recognizers-text-number-with-unit/src/main/java/com/microsoft/recognizers/text/numberwithunit/utilities/DictionaryUtils.java
package com.microsoft.recognizers.text.numberwithunit.utilities;

import com.microsoft.recognizers.text.utilities.FormatUtility;

import java.util.Map;
import java.util.regex.Pattern;

public abstract class DictionaryUtils {

    /**
     * Safely bind dictionary which contains several key-value pairs to the destination dictionary.
     * This function is used to bind all the prefix and suffix for units.
     */
    public static void bindDictionary(Map<String, String> dictionary,
                                      Map<String, String> sourceDictionary) {
        if (dictionary == null) return;

        for (Map.Entry<String, String> pair : dictionary.entrySet()) {
            if (pair.getKey() == null || pair.getKey().isEmpty()) {
                continue;
            }

            bindUnitsString(sourceDictionary, pair.getKey(), pair.getValue());
        }
    }

    /**
     * Bind keys in a string which contains words separated by '|'.
     */
    public static void bindUnitsString(Map<String, String> sourceDictionary, String key, String source) {
        String[] values = source.trim().split(Pattern.quote("|"));

        for (String token : values) {
            if (token.isEmpty() || sourceDictionary.containsKey(token)) {
                continue;
            }

            sourceDictionary.put(token, key);
        }
    }
}
